include/nbl/asset/ICPUDescriptorSet.h

```diff
@@ -144,57 +171,37 @@ class NBL_API ICPUDescriptorSet final : public IDescriptorSet<ICPUDescriptorSetL
-		}
+			const auto bindingInfo = std::lower_bound(getLayout()->getBindings().begin(), getLayout()->getBindings().end(), ICPUDescriptorSetLayout::SBinding{binding});
+			assert(bindingInfo->binding == binding && "binding is not in the descriptor set!");
 
-		//!
-		inline E_DESCRIPTOR_TYPE getDescriptorsType(uint32_t index) const
-		{
-			if (m_bindingInfo && index<m_bindingInfo->size())
-				return m_bindingInfo->operator[](index).descriptorType;
-			return EDT_INVALID;
-		}
+			const uint32_t descriptorOffset = getLayout()->getDescriptorOffset(bindingInfo->type, binding);
+			const uint32_t descriptorCount = bindingInfo->count;
 
-		//! Can modify the array of descriptors bound to a particular bindings
-		inline core::SRange<SDescriptorInfo> getDescriptors(uint32_t index) 
-		{ 
-			assert(!isImmutable_debug());
+			auto descriptorsBegin = m_descriptors[bindingInfo->type]->begin() + descriptorOffset;
+			auto descriptorInfosBegin = m_descriptorInfos[bindingInfo->type]->begin() + descriptorOffset;
 
-			if (m_bindingInfo && index<m_bindingInfo->size())
-			{
-				const auto& info = m_bindingInfo->operator[](index);
-				auto _begin = m_descriptors->begin()+info.offset;
-				if (index+1u!=m_bindingInfo->size())
-					return core::SRange<SDescriptorInfo>{_begin, m_descriptors->begin()+m_bindingInfo->operator[](index+1u).offset};
-				else
-					return core::SRange<SDescriptorInfo>{_begin, m_descriptors->end()};
-			}
-			else
-				return core::SRange<SDescriptorInfo>{nullptr, nullptr};
-		}
-		inline core::SRange<const SDescriptorInfo> getDescriptors(uint32_t index) const
-		{
-			if (m_bindingInfo && index<m_bindingInfo->size())
-			{
-				const auto& info = m_bindingInfo->operator[](index);
-				auto _begin = m_descriptors->begin()+info.offset;
-				if (index+1u!=m_bindingInfo->size())
-					return core::SRange<const SDescriptorInfo>{_begin, m_descriptors->begin()+m_bindingInfo->operator[](index+1u).offset};
-				else
-					return core::SRange<const SDescriptorInfo>{_begin, m_descriptors->end()};
-			}
-			else
-				return core::SRange<const SDescriptorInfo>{nullptr, nullptr};
+			return { {descriptorsBegin, descriptorsBegin+descriptorCount}, {descriptorInfosBegin, descriptorInfosBegin+descriptorCount} };
 		}
 
-		inline auto getTotalDescriptorCount() const
+		core::SRange<core::smart_refctd_ptr<ICPUSampler>> getMutableSamplers(const uint32_t binding) const
 		{
-			return m_descriptors->size();
+			const uint32_t offset = getLayout()->getMutableSamplerOffset(binding);
+			if (offset == ~0u)
+				return { nullptr, nullptr };
+
+			const auto bindingInfo = std::lower_bound(getLayout()->getBindings().begin(), getLayout()->getBindings().end(), ICPUDescriptorSetLayout::SBinding{ binding });
+			assert(bindingInfo->binding == binding && "binding is not in the descriptor set!");
+
+			auto samplersBegin = m_mutableSamplers->begin() + offset;
+			return { samplersBegin, samplersBegin + bindingInfo->count };
 		}
```

Comment:
we don't need those, only the infos

but you can take all the logic from them and apply them to make `getXXXXStorage(const E_DESCRIPTOR_TYPE, const binding_t)`